--- server_basic.cc	2018-05-19 15:37:53.661079635 +0800
+++ server_multiloop.cc	2018-05-20 08:58:14.543488754 +0800
@@ -18,18 +18,21 @@
 class SudokuServer
 {
  public:
-  SudokuServer(EventLoop* loop, const InetAddress& listenAddr)
+  SudokuServer(EventLoop* loop, const InetAddress& listenAddr, int numEventloops)
     : server_(loop, listenAddr, "SudokuServer"),
+      numEventloops_(numEventloops),
       startTime_(Timestamp::now())
   {
     server_.setConnectionCallback(
         std::bind(&SudokuServer::onConnection, this, _1));
     server_.setMessageCallback(
         std::bind(&SudokuServer::onMessage, this, _1, _2, _3));
+    server_.setThreadNum(numEventloops);
   }
 
   void start()
   {
+    LOG_INFO << "starting " << numEventloops_ << " threads.";
     server_.start();
   }
 
@@ -111,15 +114,21 @@
   }
 
   TcpServer server_;
+  int numEventloops_;
   Timestamp startTime_;
 };
 
 int main(int argc, char* argv[])
 {
   LOG_INFO << "pid = " << getpid() << ", tid = " << CurrentThread::tid();
+  int numEventloops = 0;
+  if (argc > 1)
+  {
+    numEventloops = atoi(argv[1]);
+  }
   EventLoop loop;
   InetAddress listenAddr(9981);
-  SudokuServer server(&loop, listenAddr);
+  SudokuServer server(&loop, listenAddr, numEventloops);
 
   server.start();
 
